x-pack/libs/es-opensaml-security-api/build.gradle

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+apply plugin: 'elasticsearch.build'
+apply plugin: 'com.github.johnrengelman.shadow'
+
+dependencies {
+  implementation "org.opensaml:opensaml-security-api:${versions.opensaml}"
+}
+
+['jarHell', 'thirdPartyAudit', 'forbiddenApisMain', 'splitPackagesAudit', 'checkstyleMain', 'licenseHeaders', 'spotlessJavaCheck'].each {
+  tasks.named(it).configure {
+    enabled = false
+  }
+}
+
+tasks.named("dependencyLicenses").configure {
+  mapping from: /opensaml-.*/, to: 'shibboleth'
+}
+
+tasks.named("shadowJar").configure {
+  manifest {
+    attributes 'Automatic-Module-Name': 'org.opensaml.security'
+  }
+  exclude 'META-INF/services/org.opensaml.security.crypto.ec.NamedCurve'
```

Comment:
Is this the only modification to this JAR we're making? Typically a "shadow" JAR is repackaging several dependencies, or relocating them under a different package, etc. Is this effectively identical to the original JAR, just excluding this file?

Reply:
Pretty much.

OpenSAML out of the box doesn't work on FIPS. My first attempt to solve it required patching 2 classes. Albert worked out that we can just avoid ever loading those classes by manipulating the set of services - one of them can be handled by stripping this service, the other by implementing our own properties resolver so that a default property can be changed.

It does seem like overkill to repackage just to drop one `services` file, but I'm not aware of a simpler option.
We could just strip the file as part of the distribution, but I'm not a fan of shipping modified files JAR files that claim to be the originals.

Reply:
Gotcha, makes sense. I would suggest we go with the pattern we used for log4j though (look in `libs/log4j/build.gradle` rather than use the Shadow plugin. It's not really the intended use case, and it's more complex than necessary for this, plus the log4j hack plays more nicely with IDEs.

Reply:
++
The original attempt modified a couple of classes, so it was based on the approach we used for Jackson.
Now that we've settled on just stripping classes, we can simplify it.

Reply:
> >    @mark-vieira I think it is possible to put the OpenSamlXpackSecurityConfigurationPropertiesSource service provider inside the shadowed jar, rather than put it in the ES code that uses the code that depends on the service. That's because the same code that uses the shadowed jar also uses the library code that depends on the service implementation.
    But I'm not sure if this is better/preferable?

> I think it's preferable to co-locate it with the shadowed jar. If for no other reason then it removes the existing duplication.

I've pushed [404298c](https://github.com/elastic/elasticsearch/pull/98199/commits/404298c1a333c5f78a779abb82b055d51814edad) and [c179601](https://github.com/elastic/elasticsearch/pull/98199/commits/c1796018818ed165e8948dbc0e6e342ca83e21fd) to incoporate the `ConfigurationPropertiesSource` service implementation in the shadowed jar, in order to avoid having two identical such implementations in the `x-pack/plugin/security` and `x-pack/plugin/identity-provider` projects (it works because the same code that must **not** see the `NamedCurve` implementations, also needs to see the new `ConfigurationPropertiesSource` implementation). So now the shadowed jar effectively removes one service implementation and adds another different one (a single new class).
I think this is now a fair usage of the "shadow" plugin?

Reply:
Yes, a Jar with somecode with bundled dependencies is more of an actual shadow jar. Although our motivation for bundling the dependency is still a bit wonky. Ideally there's a better long term plan here? Have we raised this issue with the OpenSAML folks?

Reply:
> Ideally there's a better long term plan here? Have we raised this issue with the OpenSAML folks?

That's a longer story than can be covered on this issue, but the short answer is that this is the best long term plan we have.
We've sat on the last compatible version of the dependency for some time, hoping it would be resolved, but eventually we decided we had to have the ability to move the dependency forward, even if it's an imperfect solution.

Reply:
I couldn't find discussions around bouncy castle, on the shibboleth dev archive in the past 2 yrs: http://shibboleth.net/pipermail/dev/ . Similarly there's nothing on their public Jira issue board: https://shibboleth.atlassian.net/jira/software/c/projects/OSJ/issues/?filter=allissues . The only short blurb I could find is on their wiki where they explain the usage of the bcprov jar: https://shibboleth.atlassian.net/wiki/spaces/DEV/pages/1159627167/FIPS . I can also see they're actively working on the v5 release: https://shibboleth.atlassian.net/wiki/spaces/IDP5/pages/3199511632/DevelopmentGuide

At a high level, there's a lack of true modularization across the board:
we pull in the opensaml dependency and only use a small part of its functionality, and then there's opensaml that pulls in bouncy castle of which they only use a small part of.
On the OpenSAML side I think they're fundamentally not interested in modularization because the lib is primarily designed to serve theire IDP solution.

I think we could invest into more sensible shadowing of either opensaml or bouncy castle, to obtain true modularization (eg extract only the parsing part of bcprov that opensaml uses) or even explore ways to use alternative libraries. But this is a biggish investment....

Reply:
> and then there's opensaml that pulls in bouncy castle of which they only use a small part of

Can someone explain the bouncy castle bit? When we say "pull in" what does that mean? Is it a transitive dependency or is it shadowed in the OpenSAML jar? We manually manage transitive dependencies so we can use whatever bouncy castle jar we want, or omit it entirely.

Reply:
> When we say "pull in" what does that mean? Is it a transitive dependency or is it shadowed in the OpenSAML jar?

It is a runtime transitive dependency of OpenSAML (since version >4.1).

> We manually manage transitive dependencies so we can use whatever bouncy castle jar we want, or omit it entirely.

We wish to omit it entirely because it contains crypto algorithms that are **not** FIPS compliant (there is a different bouncy castle with FIPS compliant algorithms), and that's a problem because we don't want the Elasticsearch FIPS distribution to ship with non-FIPS compliant crypto on the classpath (or really, no Elasticsearch distrib should have any crypto impl on the classpath. Instead we opt to delegate to the JCA API, where the JVM is configured for FIPS or not.).

Reply:
Ok, then I don't quite understand why we need this solution. We disable transitive dependencies by default in our build. The transitive BC jar shouldn't be brought in by this dependency. How does this relate to omitting that service file? From what I can tell that's part of the OpenSAML jar, not BC.

Reply:
If we don't put the BC on the classpath and we don't shadow jar the OpenSAML lib, then ES throws a class not found exception when it starts up, see this issue: https://github.com/elastic/elasticsearch/issues/71983 .

Reply:
Got it. So there is a _runtime_ dependency on BC and stripping out that service avoids that issue. Understood.

Reply:
Yes, OpenSAML has had a transitive dependency on BouncyCastle for a while, but we've just avoided those parts.
In 4.1 the dependency become harder to avoid - a couple of classes that are eagerly loaded (with the the default services/properties config) have a direct dependency.